# SQL Basics

## Introduction

This notebook is a short introduction to SQL. It is based on the [DuckDB](https://duckdb.org/) database engine.

## Datasets

The following datasets are used in this notebook. You don't need to download them, they can be accessed directly from the notebook.

- [cities.csv](https://storage.googleapis.com/qm2/CASA0025/cities.csv)
- [countries.csv](https://storage.googleapis.com/qm2/CASA0025/countries.csv)

## References

- [W3Schools SQL Tutorial](https://www.w3schools.com/sql)
- [DuckDB SQL Introduction](https://duckdb.org/docs/sql/introduction.html)

## Installation

Uncomment the following cell to install the required packages.

In [1]:
%pip install duckdb duckdb-engine jupysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 575.2/575.2 kB 1.7 MB/s  0:00:0036m-:--:--
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9/9 [jupysql]m8/9 [jupysql]
Note: you may need to restart the kernel to use updated packages.


## Library Import and Configuration

In [4]:
import duckdb
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Set configurations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.

In [5]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

## Connecting to DuckDB

Connect jupysql to DuckDB using a SQLAlchemy-style connection string. You may either connect to an in memory DuckDB, or a file backed db.

In [6]:
%sql duckdb:///:memory:
# %sql duckdb:///path/to/file.db

 If your SQL query is one line only, you may use the `%sql` magic command. For multi-line SQL query, you may use the `%%sql` magic command.

## Install extensions


Check available DuckDB extensions.

In [7]:
%%sql

SELECT * FROM duckdb_extensions();

,extension_name,loaded,installed,install_path,description,aliases,extension_version,install_mode,installed_from
0,autocomplete,False,False,,Adds support for autocomplete in the shell,[],,NOT_INSTALLED,
1,aws,False,False,,Provides features that depend on the AWS SDK,[],,NOT_INSTALLED,
2,azure,False,False,,Adds a filesystem abstraction for Azure blob s...,[],,NOT_INSTALLED,
3,core_functions,True,True,(BUILT-IN),Core function library,[],,STATICALLY_LINKED,
4,delta,False,False,,Adds support for Delta Lake,[],,NOT_INSTALLED,
5,ducklake,False,False,,"Adds support for DuckLake, SQL as a Lakehouse ...",[],,NOT_INSTALLED,
6,encodings,False,False,,All unicode encodings to UTF-8,[],,NOT_INSTALLED,
7,excel,False,False,,Adds support for Excel-like format strings,[],,NOT_INSTALLED,
8,fts,False,False,,Adds support for Full-Text Search Indexes,[],,NOT_INSTALLED,
9,httpfs,False,False,,Adds support for reading and writing files ove...,"[http, https, s3]",,NOT_INSTALLED,


DuckDB's [httpfs extension](https://duckdb.org/docs/extensions/httpfs) allows parquet and csv files to be queried remotely over http. This is useful for querying large datasets without having to download them locally. Let's install the extension and load the extension.

In [8]:
%%sql

INSTALL httpfs;
LOAD httpfs;

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,Success


## Read CSV

Use the `httpfs` extension to read the `cities.csv` file from the web.

In [9]:
%%sql

SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/cities.csv';

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
...,...,...,...,...,...,...
1244,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000
1245,1246,Sao Paulo,BRA,-23.55868,-46.62502,18845000
1246,1247,Sydney,AUS,-33.92001,151.18518,4630000
1247,1248,Singapore,SGP,1.29303,103.85582,5183700


In [8]:
%%sql

SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/countries.csv';

,id,Country,Alpha2_code,Alpha3_code,Numeric_code,Latitude,Longitude
0,1,Afghanistan,AF,AFG,4,33.0000,65.0
1,2,Albania,AL,ALB,8,41.0000,20.0
2,3,Algeria,DZ,DZA,12,28.0000,3.0
3,4,American Samoa,AS,ASM,16,-14.3333,-170.0
4,5,Andorra,AD,AND,20,42.5000,1.6
...,...,...,...,...,...,...,...
238,239,Wallis and Futuna,WF,WLF,876,-13.3000,-176.2
239,240,Western Sahara,EH,ESH,732,24.5000,-13.0
240,241,Yemen,YE,YEM,887,15.0000,48.0
241,242,Zambia,ZM,ZMB,894,-15.0000,30.0


## Create Table

Create a table named `cities` from the `cities.csv` file.

In [10]:
%%sql

CREATE TABLE cities AS SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/cities.csv';

,Success


Create a table named `countries` from the `countries.csv` file.

In [10]:
%%sql

CREATE TABLE countries AS SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/countries.csv';

,Success


Display the table content in the database.

In [11]:
%%sql

FROM cities;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
...,...,...,...,...,...,...
1244,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000
1245,1246,Sao Paulo,BRA,-23.55868,-46.62502,18845000
1246,1247,Sydney,AUS,-33.92001,151.18518,4630000
1247,1248,Singapore,SGP,1.29303,103.85582,5183700


In [12]:
%%sql

FROM countries;

,id,Country,Alpha2_code,Alpha3_code,Numeric_code,Latitude,Longitude
0,1,Afghanistan,AF,AFG,4,33.0000,65.0
1,2,Albania,AL,ALB,8,41.0000,20.0
2,3,Algeria,DZ,DZA,12,28.0000,3.0
3,4,American Samoa,AS,ASM,16,-14.3333,-170.0
4,5,Andorra,AD,AND,20,42.5000,1.6
...,...,...,...,...,...,...,...
238,239,Wallis and Futuna,WF,WLF,876,-13.3000,-176.2
239,240,Western Sahara,EH,ESH,732,24.5000,-13.0
240,241,Yemen,YE,YEM,887,15.0000,48.0
241,242,Zambia,ZM,ZMB,894,-15.0000,30.0


## The SQL SELECT statement

The `SELECT` statement is used to select data from a database. Use either `SELECT *` to select all columns, or `SELECT column1, column2, ...` to select specific columns.

`SELECT * FROM cities` is the same as `FROM cities`.

In [13]:
%%sql

SELECT * FROM cities;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
...,...,...,...,...,...,...
1244,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000
1245,1246,Sao Paulo,BRA,-23.55868,-46.62502,18845000
1246,1247,Sydney,AUS,-33.92001,151.18518,4630000
1247,1248,Singapore,SGP,1.29303,103.85582,5183700


To limit the number of rows returned, use the `LIMIT` keyword. For example, `SELECT * FROM cities LIMIT 10` will return only the first 10 rows.

In [14]:
%%sql

SELECT * FROM cities LIMIT 10;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
5,6,Mariehamn,ALD,60.09700,19.94900,10682
6,7,Ramallah,PSE,31.90294,35.20621,24599
7,8,Vatican City,VAT,41.90001,12.44781,832
8,9,Poitier,FRA,46.58329,0.33328,85960
9,10,Clermont-Ferrand,FRA,45.77998,3.08001,233050


Select a subset of columns from the `cities` table and display the first 10 rows.

In [15]:
%%sql

SELECT name, country FROM cities LIMIT 10;

,name,country
0,Bombo,UGA
1,Fort Portal,UGA
2,Potenza,ITA
3,Campobasso,ITA
4,Aosta,ITA
5,Mariehamn,ALD
6,Ramallah,PSE
7,Vatican City,VAT
8,Poitier,FRA
9,Clermont-Ferrand,FRA


To select distinct values, use the `DISTINCT` keyword. For example, `SELECT DISTINCT country FROM cities` will return only the distinct values of the `country` column.

In [16]:
%%sql

SELECT DISTINCT country FROM cities LIMIT 10;

,country
0,VAT
1,NLD
2,GEO
3,PRT
4,PAK
5,UZB
6,NER
7,POL
8,CHE
9,EGY


To count the number of rows returned, use the `COUNT(*)` function. For example, `SELECT COUNT(*) FROM cities` will return the number of rows in the `cities` table.

In [17]:
%%sql

SELECT COUNT(*) FROM cities;

,count_star()
0,1249


To count the number of distinct values, use the `COUNT(DISTINCT column)` function. For example, `SELECT COUNT(DISTINCT country) FROM cities` will return the number of distinct values in the `country` column.

In [18]:
%%sql

SELECT COUNT(DISTINCT country) FROM cities;

,count(DISTINCT country)
0,200


To calculate the maximum value, use the `MAX(column)` function. For example, `SELECT MAX(population) FROM cities` will return the maximum value in the `population` column.

In [19]:
%%sql

SELECT MAX(population) FROM cities;

,max(population)
0,35676000


To calculate the total value, use the `SUM(column)` function. For example, `SELECT SUM(population) FROM cities` will return the total value in the `population` column.

In [20]:
%%sql

SELECT SUM(population) FROM cities;

,sum(population)
0,1.475535e+09


To calculate the average value, use the `AVG(column)` function. For example, `SELECT AVG(population) FROM cities` will return the average value in the `population` column.

In [21]:
%%sql

SELECT AVG(population) FROM cities;

,avg(population)
0,1.181373e+06


To order the results, use the `ORDER BY column` clause. For example, `SELECT * FROM cities ORDER BY country` will return the rows ordered by the `country` column alphabetically.

In [22]:
%%sql

SELECT * FROM cities ORDER BY country LIMIT 10;

,id,name,country,latitude,longitude,population
0,665,Herat,AFG,34.33001,62.16999,481009
1,666,Mazar-e Sharif,AFG,36.69999,67.10003,458151
2,1024,Kandahar,AFG,31.61002,65.69495,715542
3,1219,Kabul,AFG,34.51669,69.18326,3277000
4,669,Malanje,AGO,-9.54000,16.34003,125856
5,670,Benguela,AGO,-12.57826,13.40723,151226
6,671,Lubango,AGO,-14.91001,13.49002,125632
7,1026,Menongue,AGO,-14.66661,17.69999,13030
8,1027,Huambo,AGO,-12.74999,15.76001,1100000
9,1177,Luanda,AGO,-8.83829,13.23443,5172900


To order the results in descending order, use the `ORDER BY column DESC` clause. For example, `SELECT * FROM cities ORDER BY country ASC, population DESC` will return the rows ordered by the `country` column alphabetical order and then by the `population` column in descending order.

In [23]:
%%sql

SELECT * FROM cities ORDER BY country ASC, population DESC LIMIT 10;

,id,name,country,latitude,longitude,population
0,1219,Kabul,AFG,34.51669,69.18326,3277000
1,1024,Kandahar,AFG,31.61002,65.69495,715542
2,665,Herat,AFG,34.33001,62.16999,481009
3,666,Mazar-e Sharif,AFG,36.69999,67.10003,458151
4,1177,Luanda,AGO,-8.83829,13.23443,5172900
5,1027,Huambo,AGO,-12.74999,15.76001,1100000
6,670,Benguela,AGO,-12.57826,13.40723,151226
7,672,Namibe,AGO,-15.19004,12.16002,132900
8,669,Malanje,AGO,-9.54000,16.34003,125856
9,671,Lubango,AGO,-14.91001,13.49002,125632


## The WHERE Clause

The `WHERE` clause is used to filter records. The `WHERE` clause is used to extract only those records that fulfill a specified condition.

In [24]:
%%sql

SELECT * FROM cities WHERE country='USA'

,id,name,country,latitude,longitude,population
0,115,Agana,USA,13.47002,144.75002,122411
1,124,San Bernardino,USA,34.12038,-117.30003,1745000
2,125,Bridgeport,USA,41.17998,-73.19996,1018000
3,126,Rochester,USA,43.17043,-77.61995,755000
4,160,International Falls,USA,48.60113,-93.41085,15240
...,...,...,...,...,...,...
109,1186,Atlanta,USA,33.83001,-84.39995,4506000
110,1187,Chicago,USA,41.82999,-87.75005,8990000
111,1223,Los Angeles,USA,33.98998,-118.17998,12500000
112,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


You can use boolean operators such as `AND`, `OR`, `NOT` to filter records. For example, `SELECT * FROM cities WHERE country='USA' OR country='CAN'` will return the rows where the `country` column is either `USA` or `CAN`.

In [25]:
%%sql

SELECT * FROM cities WHERE country='USA' OR country='CAN';

,id,name,country,latitude,longitude,population
0,115,Agana,USA,13.47002,144.75002,122411
1,124,San Bernardino,USA,34.12038,-117.30003,1745000
2,125,Bridgeport,USA,41.17998,-73.19996,1018000
3,126,Rochester,USA,43.17043,-77.61995,755000
4,160,International Falls,USA,48.60113,-93.41085,15240
...,...,...,...,...,...,...
154,1216,Vancouver,CAN,49.27342,-123.12164,2313328
155,1217,Toronto,CAN,43.69998,-79.42002,5213000
156,1223,Los Angeles,USA,33.98998,-118.17998,12500000
157,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


To select US cities with a population greater than 1 million, use the following query: `SELECT * FROM cities WHERE country='USA' AND population > 1000000`.

In [26]:
%%sql

SELECT * FROM cities WHERE country='USA' AND population>1000000;

,id,name,country,latitude,longitude,population
0,124,San Bernardino,USA,34.12038,-117.30003,1745000
1,125,Bridgeport,USA,41.17998,-73.19996,1018000
2,183,Kansas City,USA,39.10709,-94.60409,1469000
3,188,Ft. Worth,USA,32.73998,-97.34004,1440454
4,190,Austin,USA,30.26695,-97.74278,1161000
5,198,Orlando,USA,28.50998,-81.38003,1350000
6,202,Indianapolis,USA,39.74999,-86.17005,1436000
7,207,Baltimore,USA,39.29999,-76.61998,2255000
8,735,San Jose,USA,37.29998,-121.84999,1668000
9,736,Sacramento,USA,38.57502,-121.47004,1604000


To select cities with the country name starting with the letter `U`, use the following query: `SELECT * FROM cities WHERE country LIKE 'U%'`.

In [27]:
%%sql

SELECT * FROM cities WHERE country LIKE 'U%';

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,14,Jinja,UGA,0.44042,33.19993,301619
3,15,Arua,UGA,3.02037,30.90002,250000
4,16,Mbale,UGA,1.09041,34.17000,402368
...,...,...,...,...,...,...
134,1189,Kiev,UKR,50.43337,30.51663,2709000
135,1191,Tashkent,UZB,41.31170,69.29493,2184000
136,1223,Los Angeles,USA,33.98998,-118.17998,12500000
137,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


To select cities with the country name ending with the letter `A`, use the following query: `SELECT * FROM cities WHERE country LIKE '%A'`.

In [28]:
%%sql

SELECT * FROM cities WHERE country LIKE '%A';

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
...,...,...,...,...,...,...
302,1232,Lagos,NGA,6.44326,3.39153,9466000
303,1233,Rome,ITA,41.89596,12.48326,3339000
304,1242,Paris,FRA,48.86669,2.33334,9904000
305,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000


To select cities with the country name containing the letter `S` in the middle, use the following query: `SELECT * FROM cities WHERE country LIKE '_S_'`.

In [29]:
%%sql

SELECT * FROM cities WHERE country LIKE '_S_';

,id,name,country,latitude,longitude,population
0,7,Ramallah,PSE,31.90294,35.20621,24599
1,113,Palikir,FSM,6.91664,158.14997,4645
2,115,Agana,USA,13.47002,144.75002,122411
3,124,San Bernardino,USA,34.12038,-117.30003,1745000
4,125,Bridgeport,USA,41.17998,-73.19996,1018000
...,...,...,...,...,...,...
129,1187,Chicago,USA,41.82999,-87.75005,8990000
130,1192,Madrid,ESP,40.40003,-3.68335,5567000
131,1223,Los Angeles,USA,33.98998,-118.17998,12500000
132,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


To select cities from a list of countries, use the `IN` operator. For example, `SELECT * FROM cities WHERE country IN ('USA', 'CAN')` will return the rows where the `country` column is either `USA` or `CAN`.

In [30]:
%%sql

SELECT * FROM cities WHERE country IN ('USA', 'CAN');

,id,name,country,latitude,longitude,population
0,115,Agana,USA,13.47002,144.75002,122411
1,124,San Bernardino,USA,34.12038,-117.30003,1745000
2,125,Bridgeport,USA,41.17998,-73.19996,1018000
3,126,Rochester,USA,43.17043,-77.61995,755000
4,160,International Falls,USA,48.60113,-93.41085,15240
...,...,...,...,...,...,...
154,1216,Vancouver,CAN,49.27342,-123.12164,2313328
155,1217,Toronto,CAN,43.69998,-79.42002,5213000
156,1223,Los Angeles,USA,33.98998,-118.17998,12500000
157,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


To select cities with a population between 1 and 10 million, use the following query: `SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000`.

In [31]:
%%sql

SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000;

,id,name,country,latitude,longitude,population
0,98,Turin,ITA,45.07039,7.66996,1652000
1,104,Lille,FRA,50.64997,3.08001,1044000
2,124,San Bernardino,USA,34.12038,-117.30003,1745000
3,125,Bridgeport,USA,41.17998,-73.19996,1018000
4,127,Manchester,GBR,53.50042,-2.24799,2230000
...,...,...,...,...,...,...
371,1242,Paris,FRA,48.86669,2.33334,9904000
372,1243,Santiago,CHL,-33.45001,-70.66704,5720000
373,1247,Sydney,AUS,-33.92001,151.18518,4630000
374,1248,Singapore,SGP,1.29303,103.85582,5183700


## SQL Joins

Reference: https://www.w3schools.com/sql/sql_join.asp

Here are the different types of the JOINs in SQL:

- `(INNER) JOIN`: Returns records that have matching values in both tables
- `LEFT (OUTER) JOIN`: Returns all records from the left table, and the matched records from the right table
- `RIGHT (OUTER) JOIN`: Returns all records from the right table, and the matched records from the left table
- `FULL (OUTER) JOIN`: Returns all records when there is a match in either left or right table

![](https://i.imgur.com/mITYzuS.png)

We have two sample tables: `cities` and `countries`.

There are 1,249 cities in the `cities` table and 243 countries in the `countries` table.

In [32]:
%%sql

SELECT COUNT(*) FROM cities;

,count_star()
0,1249


In [33]:
%%sql

SELECT * FROM cities LIMIT 10;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
5,6,Mariehamn,ALD,60.09700,19.94900,10682
6,7,Ramallah,PSE,31.90294,35.20621,24599
7,8,Vatican City,VAT,41.90001,12.44781,832
8,9,Poitier,FRA,46.58329,0.33328,85960
9,10,Clermont-Ferrand,FRA,45.77998,3.08001,233050


In [34]:
%%sql

SELECT COUNT(*) FROM countries;

,count_star()
0,243


In [35]:
%%sql

SELECT * FROM countries LIMIT 10;

,id,Country,Alpha2_code,Alpha3_code,Numeric_code,Latitude,Longitude
0,1,Afghanistan,AF,AFG,4,33.0000,65.0000
1,2,Albania,AL,ALB,8,41.0000,20.0000
2,3,Algeria,DZ,DZA,12,28.0000,3.0000
3,4,American Samoa,AS,ASM,16,-14.3333,-170.0000
4,5,Andorra,AD,AND,20,42.5000,1.6000
5,6,Angola,AO,AGO,24,-12.5000,18.5000
6,7,Anguilla,AI,AIA,660,18.2500,-63.1667
7,8,Antarctica,AQ,ATA,10,-90.0000,0.0000
8,9,Antigua and Barbuda,AG,ATG,28,17.0500,-61.8000
9,10,Argentina,AR,ARG,32,-34.0000,-64.0000


### SQL Inner Join

The `INNER JOIN` keyword selects records that have matching values in both tables. In the example, we join the `cities` table with the `countries` table using the `country` column in the `cities` table and the `Alpha3_code` column in the `countries` table. The result contains 1,244 rows, indicating that there are 5 cities that do not have a matching country.

In [36]:
%%sql

SELECT * FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

,id,name,country,latitude,longitude,population,id_1,Country_1,Alpha2_code,Alpha3_code,Numeric_code,Latitude_1,Longitude_1
0,1,Bombo,UGA,0.58330,32.53330,75000,226,Uganda,UG,UGA,800,1.0000,32.0000
1,2,Fort Portal,UGA,0.67100,30.27500,42670,226,Uganda,UG,UGA,800,1.0000,32.0000
2,3,Potenza,ITA,40.64200,15.79900,69060,106,Italy,IT,ITA,380,42.8333,12.8333
3,4,Campobasso,ITA,41.56300,14.65600,50762,106,Italy,IT,ITA,380,42.8333,12.8333
4,5,Aosta,ITA,45.73700,7.31500,34062,106,Italy,IT,ITA,380,42.8333,12.8333
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1239,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000,30,Brazil,BR,BRA,76,-10.0000,-55.0000
1240,1246,Sao Paulo,BRA,-23.55868,-46.62502,18845000,30,Brazil,BR,BRA,76,-10.0000,-55.0000
1241,1247,Sydney,AUS,-33.92001,151.18518,4630000,13,Australia,AU,AUS,36,-27.0000,133.0000
1242,1248,Singapore,SGP,1.29303,103.85582,5183700,195,Singapore,SG,SGP,702,1.3667,103.8000


Only select the `city` and `country` columns from the `cities` table and the `Country` column from the `countries` table.

In [37]:
%%sql

SELECT name, cities."country", countries."Country" FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

,name,country,Country_1
0,Bombo,UGA,Uganda
1,Fort Portal,UGA,Uganda
2,Potenza,ITA,Italy
3,Campobasso,ITA,Italy
4,Aosta,ITA,Italy
...,...,...,...
1239,Rio de Janeiro,BRA,Brazil
1240,Sao Paulo,BRA,Brazil
1241,Sydney,AUS,Australia
1242,Singapore,SGP,Singapore


### SQL Left Join

The `LEFT JOIN` keyword returns all records from the left table (`cities`), and the matched records from the right table (`countries`). The result contains 1,249 rows, the same number of rows as the `cities` table.

In [38]:
%%sql

SELECT * FROM cities LEFT JOIN countries ON cities.country = countries."Alpha3_code";

,id,name,country,latitude,longitude,population,id_1,Country_1,Alpha2_code,Alpha3_code,Numeric_code,Latitude_1,Longitude_1
0,1,Bombo,UGA,0.58330,32.53330,75000,226,Uganda,UG,UGA,800,1.0000,32.0000
1,2,Fort Portal,UGA,0.67100,30.27500,42670,226,Uganda,UG,UGA,800,1.0000,32.0000
2,3,Potenza,ITA,40.64200,15.79900,69060,106,Italy,IT,ITA,380,42.8333,12.8333
3,4,Campobasso,ITA,41.56300,14.65600,50762,106,Italy,IT,ITA,380,42.8333,12.8333
4,5,Aosta,ITA,45.73700,7.31500,34062,106,Italy,IT,ITA,380,42.8333,12.8333
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1244,6,Mariehamn,ALD,60.09700,19.94900,10682,<NA>,None,None,None,<NA>,NaN,NaN
1245,118,Bir Lehlou,SAH,26.11917,-9.65252,500,<NA>,None,None,None,<NA>,NaN,NaN
1246,415,Pristina,KOS,42.66671,21.16598,465186,<NA>,None,None,None,<NA>,NaN,NaN
1247,719,Berbera,SOL,10.43555,45.01641,242344,<NA>,None,None,None,<NA>,NaN,NaN


### SQL Right Join

The `RIGHT JOIN` keyword returns all records from the right table (`countries`), and the matched records from the left table (`cities`). The result contains 1,291 rows.

In [39]:
%%sql

SELECT * FROM cities RIGHT JOIN countries ON cities.country = countries."Alpha3_code";

,id,name,country,latitude,longitude,population,id_1,Country_1,Alpha2_code,Alpha3_code,Numeric_code,Latitude_1,Longitude_1
0,1,Bombo,UGA,0.5833,32.5333,75000,226,Uganda,UG,UGA,800,1.0000,32.0000
1,2,Fort Portal,UGA,0.6710,30.2750,42670,226,Uganda,UG,UGA,800,1.0000,32.0000
2,3,Potenza,ITA,40.6420,15.7990,69060,106,Italy,IT,ITA,380,42.8333,12.8333
3,4,Campobasso,ITA,41.5630,14.6560,50762,106,Italy,IT,ITA,380,42.8333,12.8333
4,5,Aosta,ITA,45.7370,7.3150,34062,106,Italy,IT,ITA,380,42.8333,12.8333
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1286,<NA>,None,None,NaN,NaN,<NA>,224,Turks and Caicos Islands,TC,TCA,796,21.7500,-71.5833
1287,<NA>,None,None,NaN,NaN,<NA>,88,Guernsey,GG,GGY,831,49.5000,-2.5600
1288,<NA>,None,None,NaN,NaN,<NA>,136,Martinique,MQ,MTQ,474,14.6667,-61.0000
1289,<NA>,None,None,NaN,NaN,<NA>,139,Mayotte,YT,MYT,175,-12.8333,45.1667


### SQL Full Join

The `FULL JOIN` keyword returns all records when there is a match in either left or right table. The result contains 1,296 rows.

In [40]:
%%sql

SELECT * FROM cities FULL JOIN countries ON cities.country = countries."Alpha3_code";

,id,name,country,latitude,longitude,population,id_1,Country_1,Alpha2_code,Alpha3_code,Numeric_code,Latitude_1,Longitude_1
0,1,Bombo,UGA,0.5833,32.5333,75000,226,Uganda,UG,UGA,800,1.0000,32.0000
1,2,Fort Portal,UGA,0.6710,30.2750,42670,226,Uganda,UG,UGA,800,1.0000,32.0000
2,3,Potenza,ITA,40.6420,15.7990,69060,106,Italy,IT,ITA,380,42.8333,12.8333
3,4,Campobasso,ITA,41.5630,14.6560,50762,106,Italy,IT,ITA,380,42.8333,12.8333
4,5,Aosta,ITA,45.7370,7.3150,34062,106,Italy,IT,ITA,380,42.8333,12.8333
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1291,<NA>,None,None,NaN,NaN,<NA>,224,Turks and Caicos Islands,TC,TCA,796,21.7500,-71.5833
1292,<NA>,None,None,NaN,NaN,<NA>,88,Guernsey,GG,GGY,831,49.5000,-2.5600
1293,<NA>,None,None,NaN,NaN,<NA>,136,Martinique,MQ,MTQ,474,14.6667,-61.0000
1294,<NA>,None,None,NaN,NaN,<NA>,139,Mayotte,YT,MYT,175,-12.8333,45.1667


### SQL Union

The `UNION` operator is used to combine the result-set of two or more `SELECT` statements.

In [41]:
%%sql

SELECT country FROM cities
UNION
SELECT "Alpha3_code" FROM countries;

,country
0,PSE
1,ZMB
2,JPN
3,NGA
4,TGO
...,...
242,FRO
243,HKG
244,NRU
245,NIU


## Aggregation

### Group By

The `GROUP BY` statement groups rows that have the same values into summary rows, like "find the number of cities in each country".

The `GROUP BY` statement is often used with aggregate functions (`COUNT`, `MAX`, `MIN`, `SUM`, `AVG`) to group the result-set by one or more columns.

In [42]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
ORDER BY COUNT(name) DESC;

,"count(""name"")",country
0,114,USA
1,100,CHN
2,81,RUS
3,69,IND
4,46,BRA
...,...,...
195,1,LVA
196,1,LSO
197,1,EST
198,1,BGR


In [43]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
ORDER BY COUNT(name) DESC;

,Country,"count(""name"")"
0,United States,114
1,China,100
2,Russia,81
3,India,69
4,Brazil,46
...,...,...
192,Croatia,1
193,Fiji,1
194,Seychelles,1
195,Cyprus,1


### Having

The `HAVING` clause was added to SQL because the `WHERE` keyword could not be used with aggregate functions.

For example, to select countries with more than 40 cities:

In [44]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

,"count(""name"")",country
0,114,USA
1,100,CHN
2,81,RUS
3,69,IND
4,46,BRA
5,45,CAN


In [45]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

,Country,"count(""name"")"
0,United States,114
1,China,100
2,Russia,81
3,India,69
4,Brazil,46
5,Canada,45


## Conditional statements

The `CASE` statement goes through conditions and returns a value when the first condition is met (like an `IF-THEN-ELSE` statement). So, once a condition is true, it will stop reading and return the result. If no conditions are true, it returns the value in the `ELSE` clause.

For example, to divide cities into 3 groups based on their population:

In [46]:
%%sql

SELECT name, population,
CASE
    WHEN population > 10000000 THEN 'Megacity'
    WHEN population > 1000000 THEN 'Large city'
    ELSE 'Small city'
END AS category
FROM cities;

,name,population,category
0,Bombo,75000,Small city
1,Fort Portal,42670,Small city
2,Potenza,69060,Small city
3,Campobasso,50762,Small city
4,Aosta,34062,Small city
...,...,...,...
1244,Rio de Janeiro,11748000,Megacity
1245,Sao Paulo,18845000,Megacity
1246,Sydney,4630000,Large city
1247,Singapore,5183700,Large city


## Saving results

You can save the results of a query to a new table using the `CREATE TABLE AS` statement.

In [47]:
%%sql

CREATE TABLE cities2 AS SELECT * FROM cities;

,Success


Show the new table content.

In [48]:
%%sql

FROM cities2;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
...,...,...,...,...,...,...
1244,1245,Rio de Janeiro,BRA,-22.92502,-43.22502,11748000
1245,1246,Sao Paulo,BRA,-23.55868,-46.62502,18845000
1246,1247,Sydney,AUS,-33.92001,151.18518,4630000
1247,1248,Singapore,SGP,1.29303,103.85582,5183700


Use the `DROP TABLE` statement to delete the table.

In [49]:
%%sql

DROP TABLE IF EXISTS cities_usa;
CREATE TABLE cities_usa AS (SELECT * FROM cities WHERE country = 'USA');

,Success


In [50]:
%%sql

FROM cities_usa;

,id,name,country,latitude,longitude,population
0,115,Agana,USA,13.47002,144.75002,122411
1,124,San Bernardino,USA,34.12038,-117.30003,1745000
2,125,Bridgeport,USA,41.17998,-73.19996,1018000
3,126,Rochester,USA,43.17043,-77.61995,755000
4,160,International Falls,USA,48.60113,-93.41085,15240
...,...,...,...,...,...,...
109,1186,Atlanta,USA,33.83001,-84.39995,4506000
110,1187,Chicago,USA,41.82999,-87.75005,8990000
111,1223,Los Angeles,USA,33.98998,-118.17998,12500000
112,1224,Washington D.C.,USA,38.89955,-77.00942,4338000


Use the `INSERT INTO` statement to insert rows into a table.

In [51]:
%%sql

INSERT INTO cities_usa (SELECT * FROM cities WHERE country = 'CAN');

,Success


## SQL Comments

Comments are used to explain sections of SQL statements, or to prevent execution of SQL statements.


### Single line comMents

Single line comments start with --.

Any text between -- and the end of the line will be ignored (will not be executed).

The following example uses a single-line comment as an explanation:

In [52]:
%%sql

SELECT * FROM cities LIMIT 10 -- This is a comment;

,id,name,country,latitude,longitude,population
0,1,Bombo,UGA,0.58330,32.53330,75000
1,2,Fort Portal,UGA,0.67100,30.27500,42670
2,3,Potenza,ITA,40.64200,15.79900,69060
3,4,Campobasso,ITA,41.56300,14.65600,50762
4,5,Aosta,ITA,45.73700,7.31500,34062
5,6,Mariehamn,ALD,60.09700,19.94900,10682
6,7,Ramallah,PSE,31.90294,35.20621,24599
7,8,Vatican City,VAT,41.90001,12.44781,832
8,9,Poitier,FRA,46.58329,0.33328,85960
9,10,Clermont-Ferrand,FRA,45.77998,3.08001,233050


### Multi-line comments

Multi-line comments start with `/*` and end with `*/`.

Any text between `/*` and `*/` will be ignored.

The following example uses a multi-line comment as an explanation:

In [53]:
%%sql

SELECT COUNT(name), country
FROM cities
/*
 * Adding Group by
 * Adding Order by
 */
GROUP BY country
ORDER BY COUNT(name) DESC
LIMIT 10;

,"count(""name"")",country
0,114,USA
1,100,CHN
2,81,RUS
3,69,IND
4,46,BRA
5,45,CAN
6,40,ATA
7,36,AUS
8,30,FRA
9,27,MEX
